Allowing getting posts by ids in GET /posts response using include array

paulopmt1 opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/wp-calypso/pull/73760

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* We're adding the ability to get /posts by their IDs, so we can easily load a chunk of posts on Calypso

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Apply this code to your sandbox. You can use the `Connecting to your Sandbox` notes at: pdWQjU-77-p2#example-use-case
* Apply [these changes](https://github.com/Automattic/wp-calypso/pull/73760) to your local calypso
* Access a popular site you have and make sure we're listing posts and pages by their views at the top of /advertising items
* Tip: Comment [these lines](https://github.com/Automattic/wp-calypso/blob/update/filter-advertising-by-views-v2/client/my-sites/promote-post/main.tsx#L122-L143) and access any private site from automattic that have more views, like the explorers p2.
* You should be able to see /advertising posts and note that we're loading a chunck of them by calling `https://public-api.wordpress.com/rest/v1.1/sites/163795312/posts?http_envelope=1&include%5B%5D=0&include%5B%5D=8671...`

We can also validate it directly by calling the API endpoint passing the posts_ids we want to load, example `/sites/163795312/posts?http_envelope=1&include[]=8671&include[]=9310&include[]=9794`:

<img width="881" alt="image" src="https://user-images.githubusercontent.com/1044309/221189403-1e785d80-f596-4743-95d4-176e1f9c4223.png">

paulopmt1 committed:
Allowing getting posts by ids at GET /posts response

paulopmt1 committed:
changelog

lupus2k reviewed (approved):
LGTM!

jeherve commented on projects/plugins/jetpack/json-endpoints/class.wpcom-json-api-list-posts-v1-1-endpoint.php:
```diff
@@ -220,6 +221,10 @@ public function callback( $path = '', $blog_id = 0 ) {
 		}
 
+		if ( isset( $args['include'] ) ) {
+			$query['post__in'] = $args['include'];
```
Should we worry about other parameters that may already set `post__in`, such as https://github.com/Automattic/jetpack/blob/1a7961cde8fa002e29395fa181123801fff4b19a/projects/plugins/jetpack/json-endpoints/class.wpcom-json-api-list-posts-v1-1-endpoint.php#L251-L254

Maybe we should merge the 2, like we do with `post__not_in` elsewhere in this file?

paulopmt1 replied:
Good catch! [Fixed it here](https://github.com/Automattic/jetpack/pull/29149/commits/22278727539e13044e1ae3dc7cdc327fa1906060).

jeherve commented on projects/plugins/jetpack/json-endpoints/class.wpcom-json-api-list-posts-v1-1-endpoint.php:
```diff
@@ -220,6 +221,10 @@ public function callback( $path = '', $blog_id = 0 ) {
 		}
 
+		if ( isset( $args['include'] ) ) {
+			$query['post__in'] = $args['include'];
```
Do we need to do any kind of sanitizing / casting here, to ensure that we'll have the proper data depending on what was passed (array or int)?

paulopmt1 replied:
That's a good idea! [Fixed it here](https://github.com/Automattic/jetpack/pull/29149/commits/22278727539e13044e1ae3dc7cdc327fa1906060).

jeherve reviewed (commented):
I left a couple of questions below. Beyond those questions, I think it would be worth adding those changes to the 2 other versions of the endpoint (v1 and v1.2), for consistency.

paulopmt1 committed:
Added casting to include variable and improved its compatibilty

paulopmt1 committed:
Added the new filter for both v1 and v1.2 versions of the API

paulopmt1 commented:
> I left a couple of questions below. Beyond those questions, I think it would be worth adding those changes to the 2 other versions of the endpoint (v1 and v1.2), for consistency.

Hey @jeherve, thanks for your review and suggestions. Changed the casting and added that filter to the v1 and v1.2 versions of the endpoint.
